src/hotspot/share/utilities/lockFreeQueue.inline.hpp

```diff
@@ -29,45 +29,101 @@
-    }
+template<typename T, T* volatile* (*next_ptr)(T&)>
+T* LockFreeQueue<T, next_ptr>::next(const T& node) {
+  return Atomic::load(next_ptr(const_cast<T&>(node)));
+}
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+void LockFreeQueue<T, next_ptr>::set_next(T& node, T* new_next) {
+    Atomic::store(next_ptr(node), new_next);
+}
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+LockFreeQueue<T, next_ptr>::LockFreeQueue() : _head(NULL), _tail(NULL) {}
+
+#ifdef ASSERT
+template<typename T, T* volatile* (*next_ptr)(T&)>
+LockFreeQueue<T, next_ptr>::~LockFreeQueue() {
+  assert(_head == NULL, "precondition");
+  assert(_tail == NULL, "precondition");
+}
+#endif
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+T* LockFreeQueue<T, next_ptr>::top() const {
+  return Atomic::load(&_head);
+}
+
+template<typename T, T* volatile* (*next_ptr)(T&)>
+size_t LockFreeQueue<T, next_ptr>::length() const {
+  size_t result = 0;
+  for (const T* current = top(); current != NULL; current = next(*current)) {
+    ++result;
   }
-  inline ~LockFreeQueueCriticalSection() {
-    if (enable) {
-      GlobalCounter::critical_section_end(_thread, _context);
-    }
+  return result;
+}
+
+// An append operation atomically exchanges the new tail with the queue tail.
+// It then sets the "next" value of the old tail to the head of the list being
+// appended; it is an invariant that the old tail's "next" value is NULL.
+// But if the old tail is NULL then the queue was empty.  In this case the
+// head of the list being appended is instead stored in the queue head; it is
+// an invariant that the queue head is NULL in this case.
+//
+// This means there is a period between the exchange and the old tail update
+// where the queue sequence is split into two parts, the list from the queue
+// head to the old tail, and the list being appended.  If there are concurrent
+// push/append operations, each may introduce another such segment.  But they
+// all eventually get resolved by their respective updates of their old tail's
+// "next" value.  This also means that pop operations must handle an object
+// with a NULL "next" value specially.
+//
+// A push operation is just a degenerate append, where the object being pushed
+// is both the head and the tail of the list being appended.
+template<typename T, T* volatile* (*next_ptr)(T&)>
+void LockFreeQueue<T, next_ptr>::append(T& first, T& last) {
+  assert(next(last) == NULL, "precondition");
+  T* old_tail = Atomic::xchg(&_tail, &last);
+  if (old_tail == NULL) {       // Was empty.
+    Atomic::store(&_head, &first);
+  } else {
+    assert(next(*old_tail) == NULL, "invariant");
+    set_next(*old_tail, &first);
   }
-};
+}
 
-template<typename T, T* volatile* (*next_ptr)(T&), bool rcu_pop>
-T* LockFreeQueue<T, next_ptr, rcu_pop>::pop() {
-  Thread* current_thread = Thread::current();
+template<typename T, T* volatile* (*next_ptr)(T&)>
+template<bool use_rcu>
+T* LockFreeQueue<T, next_ptr>::pop() {
```

Comment:
Thanks for the suggestion, and yes this makes sense. Done. Could you double-check if the updated comments are appropriate?

My only concern is that the difference between operation_in_progress and lost_race may be too subtle for most client code. I suppose most client code can just do "retry until succeeded" like in the test.
I don't expect these two cases to differ much in run-time. Is there any performance data to show that the operation_in_progress case indeed takes much longer for retrying? I checked review threads for JDK-8237143 and JDK-8238867 but didn't find any.

Anyway, this CR should use the tri-status approach (and the intra-loop critical section approach), in order to avoid behavior change.

Reply:
In the "lost-race" cases, there was cmpxchg contention that the current thread lost. Retry is a reasonable thing to do, though that's up to the application. Success on retry is of course not guaranteed, because there could again be contention with some other thread, but at least somebody is making progress.

In the "operation-in-progress" cases the queue is in a state where the current operation cannot make progress until a different thread changes the state. If that other thread happens to have been descheduled or something like that, it could be a (comparatively) very long time before that happens, with lots of expensive spinning by a retrying try_pop.

So I think the comment for that state needs some more work.  Maybe something like this?

"An in-progress concurrent operation interfered with taking what had been the only remaining element in the queue. A concurrent try_pop may have already claimed it, but not completely updated the queue.  Alternatively, a concurrent push/append may have not yet linked the new entry(s) to the former sole entry. Retrying the try_pop will continue to fail in this way until that other thread has updated the queue's internal structure."

That was sufficient for the G1DCQS usage, so I stopped trying to do better.  (I don't even know if it's possible to do better, though I have some old notes on an idea.) But it's pretty ugly for a general utility, which is why I left this queue class buried inside G1DCQS rather than hoisting it out and generalizing it.